File: BaseController.m
% BaseController - Abstract class for implementing control systems
%
% This abstract class serves as a base class for implementing control systems.
% It includes properties and abstract methods that should be implemented
% by derived controller classes.
%
% Properties (Access = protected):
%   - Tc: Sampling time (cycle time)
%   - umax: Maximum control action
%
% Methods (Access = public):
%   - BaseController: Constructor to create BaseController objects
%   - setUMax: Method to set the maximum control action
%   - getSamplingPeriod: Method to get the sampling period (Tc)
%
% Abstract Methods (Access = public):
%   - initialize: Abstract method to (re)initialize the controller
%   - starting: Abstract method to start the controller with bumpless transition
%   - computeControlAction: Abstract method to compute the control action
%
classdef (Abstract) BaseController < handle
    properties (Access = protected)
        Tc % Sampling time (cycle time)
        umax % Maximum control action
    end

    methods
        % Constructor for BaseController class
        function obj = BaseController(Tc)
            % Assign the sampling time to the property
            obj.Tc = Tc;
        end

        % Method to set the maximum control action
        function setUMax(obj, umax)
            obj.umax = umax;
        end
        % Method to set the maximum control action
        function umax=getUMax(obj)
            umax=obj.umax;
        end
        % Method to get the sampling period
        function st = getSamplingPeriod(obj)
            st = obj.Tc;
        end
    end

    methods (Abstract, Access = public)
        % Abstract methods to be implemented in derived classes

        % Method to (re)initialize the controller
        obj = initialize(obj)

        % Method to start the controller with bumpless transition
        %   reference = setpoint
        %   y = output
        %   u = control action
        %   uff = feedforward action
        obj = starting(obj, reference, y, u, uff)

        % Method to compute the control action
        %   reference = setpoint
        %   y = output
        %   u = control action
        %   uff = feedforward action
        u = computeControlAction(obj, reference, y, uff)
    end
end

File: base_controller.py
from abc import ABC, abstractmethod

class BaseController(ABC):
    """
    BaseController - Abstract class for implementing control systems.
    
    This abstract class serves as a base class for implementing control systems.
    It includes properties and abstract methods that should be implemented
    by derived controller classes.
    """
    
    def __init__(self, Tc: float):
        """
        Constructor to create BaseController objects.
        
        :param Tc: Sampling time (cycle time), must be a positive scalar.
        """
        if not isinstance(Tc, (int, float)) or Tc <= 0:
            raise ValueError("Sampling time (Tc) must be a positive scalar")
        
        self._Tc = Tc  # Protected attribute for sampling time
        self._umax = float('inf') # Maximum control action (initialized as infinit)
    
    def set_umax(self, umax: float):
        """Set the maximum control action."""
        if not isinstance(umax, (int, float)):
            raise ValueError("Maximum control action (umax) must be a scalar")
        self._umax = umax
    
    def get_umax(self) -> float:
        """Get the maximum control action."""
        return self._umax
    
    def get_sampling_period(self) -> float:
        """Get the sampling period (Tc)."""
        return self._Tc
    
    @abstractmethod
    def initialize(self):
        """Abstract method to (re)initialize the controller."""
        pass
    
    @abstractmethod
    def starting(self, reference: float, y: float, u: float, uff: float):
        """
        Abstract method to start the controller with a bumpless transition.
        
        :param reference: Setpoint.
        :param y: Output.
        :param u: Control action.
        :param uff: Feedforward action.
        """
        pass
    
    @abstractmethod
    def compute_control_action(self, reference: float, y: float, uff: float) -> float:
        """
        Abstract method to compute the control action.
        
        :param reference: Setpoint.
        :param y: Output.
        :param uff: Feedforward action.
        :return: Computed control action.
        """
        pass
